### DISCUSSION TITLE
Adding Fleet Telemetry

### DISCUSSION THREAD

**Discussion by abreingan** [llamafilm/tesla-http-proxy-addon#107] [Q&A] [answered]
I have been trying to add a Telemetry server using a description at https://dev.to/patrickdemers6/fleet-telemetry-streaming-data-from-your-tesla-2i97.
That author suggests that the same private key can be used (should be used?) for both the command api and the telemetry. Given that I am reluctant to try generating another key in case I break the command api. 
I can see the public key in home assistant, but I cannot see the private key. Is it available somewhere?
Thanks in advance

**Comment by llamafilm** [comment]
You should be able to see it with `pass show myself` inside the container. (To get into the container, you need the advanced SSH add on to enable port 22222 and then use `docker exec`.)

**Reply by abreingan** [reply]
Thank you for a quick and helpful answer as always.
But I am afraid I am working "beyond my pay-grade" and need more advice as I think I need to import the private key to gpg and that's a permanent change that I am too cowardly to do without advice.
I did as you suggested and ended up with:-
 docker exec -e PASSWORD_STORE_DIR="/data/password-store" addon_c03d64a7_tesla_http_proxy pass show vehicleCommandKey.myself  
That fails with :-
gpg: public key decryption failed: No secret key
gpg: decryption failed: No secret key
Googling a few similar issues it seems I could "gpg --import" the private key which I assume is  /data/key.pem but before I go there it seems sensible to stop and cry help.

**Reply by llamafilm** [reply]
You almost have it, just need to add `-e GNUPGHOME=/data/gnugpg`.  That will show you the private key, generated by `tesla-keygen`, but I don't know what format it's in or how to use it elsewhere.

**Reply by abreingan** [reply]
Thanks for that which got me to the next stage, but I am afraid I am more confused than ever.
I have obviously missed something, so I start again from the front.
Both the description of how to set up the Fleet API (at https://shankarkumarasamy.blog/2023/10/29/tesla-developer-api-guide-account-setup-app-creation-registration-and-third-party-authentication-configuration-part-1/) and the description of how to set up the Telemetry (at https://dev.to/patrickdemers6/fleet-telemetry-streaming-data-from-your-tesla-2i97)  both start by generating a private key via an identical command (other than the output file names)
openssl ecparam -name prime256v1 -genkey -noout -out private-key.pem

They both then use that key to generate a public key that is stored in /.well-known/appspecific/com.tesla.3p.public-key.pem

For telemetry that original private key is used to create a CSR, and for the Fleet API I assume you continue by getting a partner authentication token etc.

Its that original private key that I am after as I need it to generate a CSR. 

I had a look through the files in /data but couldn't find anything that corresponds to /.well-known/appspecific/com.tesla.3p.public-key.pem.

Please tell me I am missing something.

Thanks again for your patience
Al

**Reply by llamafilm** [reply]
I don't know why those guides use `openssl` to create the public key.  I use `tesla-keygen` based on instructions from here: https://github.com/teslamotors/vehicle-command.

With this container, you can get the private key using `docker exec -e GNUPGHOME=/data/gnugpg -e PASSWORD_STORE_DIR="/data/password-store" addon_c03d64a7_tesla_http_proxy pass show vehicleCommandKey.myself`.

The certificate and key files in `/data` have nothing to do with this; they are used for HTTPS in `tesla-http-proxy`.

**Reply by abreingan** [reply]
Sorry Elliott, I had tried that but thought it must be something different as it was in a format I wasn't expecting. The data string returned is 44 bytes long starting with 14CI and ending in qu8= with all characters looking like ASCII which I guess is another format. I  will go dig around there to see if I can convert that into something that can be used for the CSR, though I have little experience in this area. 
Thanks again for your help

**Reply by abreingan** [reply]
Hi Elliott,
I notice that /teslamotors/vehicle-command now has an option to export the private key (I asked if such was possible). Is there any chance of getting a way to do this from your proxy?
Thanks again
Alastair

**Reply by llamafilm** [reply, marked as answer]
I've just updated to the latest tesla-keygen so this should work for you in v2.2.8.

```
docker exec \
-e GNUPGHOME=/data/gnugpg \
-e PASSWORD_STORE_DIR="/data/password-store" \
addon_c03d64a7_tesla_http_proxy tesla-keygen -keyring-type pass -key-name myself export
```

**Reply by abreingan** [reply]
Thank you again Elliott, that seemed to work perfectly.
